_posts/2024-10-01-understanding-index-request-cache.markdown

```diff
@@ -0,0 +1,158 @@
+---
+layout: post
+title: "Optimizing query performance: The implementation of index request cache in OpenSearch" 
+authors:
+   - awskiran
+   - kkhatua
+   - upasagar
+   - kolchfa
+date: 2024-10-01 
+categories:
+  - technical-posts
+  - search
+meta_keywords: OpenSearch cluster, Opensearch caching, index request cache, search performance optimization, search latency
+meta_description: 
+---
+
+Speed and efficiency are essential for search users. OpenSearch achieves these through various mechanisms, one of the most important of which is [index request cache](https://opensearch.org/docs/latest/search-plugins/caching/request-cache/). This blog post describes how this cache is implemented and how it optimizes query performance.
+
+## What is index request cache?
+
+Index request cache is designed to accelerate search queries in OpenSearch by storing query results at the shard level, as shown in the following image. 
+
+![Index-Request-Cache](/assets/media/blog-images/2024-10-01-understanding-index-request-cache/cache_location.png){:class="img-centered" style="width:750px;"}
+
+Caching frequently run queries reduces the need to re-execute the same query multiple times. This approach is particularly effective for queries targeting specific indexes or patterns, improving response times and overall system efficiency. The cache automatically clears entries when the data changes, ensuring that only up-to-date information is returned.
+
+## Caching policy
+
+Not all search requests are eligible for caching in index request cache. Search requests which specify `size=0` are cached by default. These requests cache only metadata, such as the total number of results or hits. 
+
+The following requests are ineligible for caching:
+
+* **Non-deterministic requests:** Searches involving functions like `Math.random()` or relative times such as `now` or `new Date()`.
+* **Scroll and Profile API requests**
+* **DFS query then fetch requests:** This search type depends on both index content and overridden statistics, leading to inaccurate scores when the statistics differ (for example, because of shard updates).
+
+You can enable caching for individual search requests by setting the `request_cache` query parameter to `true`:
+
+```json
+GET /students/_search?request_cache=true
+{
+  "query": {
+    "match": {
+      "name": "doe john"
+    }
+  }
+}
+```
+
+## Understanding cache entries
+
+Each cache entry is a key-value pair of **Key → BytesReference**.
+
+A [Key](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L346) consists of three entities:
```

Comment:
I think this should be a `Key` instead of `key` because i want the reader to think of this as a java class which follows the standard of camelcase and its easier to relate with the diagram that also is in the same format.

Reply:
@kiranprakash154 Ok, that works for me.

Reply:
Yes, should be capital. Classes in Java are capitalized.